Reimplement grouping as analysis input - 2

ChiCheng45 opened the pull request:
**Description of work**
Grouping has now been implemented and will be applied to all jobs that use the weights configurator and other places where it makes sense to have grouping.

**Fixes**
closes #644
closes #680
Fixed intramolecular TCF result. See typo in protos below.
https://github.com/ISISNeutronMuon/MDANSE/blob/d9f00436b13464b70d6a73e1103eaab5efa9c801/MDANSE/Src/MDANSE/Framework/Jobs/PairDistributionFunction.py#L160-L172

**Explanation**
Here is a short explanation of how grouping works with the weight scheme. The total intermediate scattering function is

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = \frac{1}{N} \sum_{j} w_j \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle 
```

when we use the atom grouping, the above sum is split up by atom.

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = \sum_{\alpha} W_{\alpha} F_{\mathrm{inc}, \alpha}(q, t) 
```

```math
F_{\mathrm{inc}, \alpha}(q, t) = \frac{1}{Nc_{\alpha}} \sum_{j \in \alpha} \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle 
```

For the molecular case, we split the summation into atomic parts associated with each molecule. For a water and ethanol system, we would get 

```math
 F_{\mathrm{inc}, \mathrm{tot}}(q, t) = c_{\mathrm{H2O}}F_{\mathrm{inc}, \mathrm{H2O}}(q, t) +c_{\mathrm{EtOH}} F_{\mathrm{inc}, \mathrm{EtOH}}(q, t) 
```

where

```math
c_{\mathrm{H2O}} F_{\mathrm{inc}, \mathrm{H2O}}(q, t) = W^{\mathrm{H2O}}_{\mathrm{H}}F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{H2O}}(q, t) + W^{\mathrm{H2O}}_{\mathrm{O}}F_{\mathrm{inc}, \mathrm{O}}^{\mathrm{H2O}}(q, t)
```

and

```math
c_{\mathrm{EtOH}}F_{\mathrm{inc}, \mathrm{EtOH}}(q, t) = W^{\mathrm{EtOH}}_{\mathrm{H}}F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{EtOH}}(q, t) + W^{\mathrm{EtOH}}_{\mathrm{C}}F_{\mathrm{inc}, \mathrm{C}}^{\mathrm{EtOH}}(q, t) + W^{\mathrm{EtOH}}_{\mathrm{O}}F_{\mathrm{inc}, \mathrm{O}}^{\mathrm{EtOH}}(q, t)
```

The intermediate scattering functions are

```math
F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{H2O}}(q, t) = \frac{1}{Nc^{\mathrm{H2O}}_{\mathrm{H}}} \sum_{j \in (\mathrm{H2O} \cap \mathrm{H})} \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle
```
and 

```math
W^{\mathrm{H2O}}_{\mathrm{H}} = c_{\mathrm{H}}^{\mathrm{H2O}} w_{\mathrm{H}}
```

where $c_{\mathrm{H2O}} = N_{\mathrm{H2O}} / N$, $c_{\mathrm{H}}^{\mathrm{H2O}} = N_{\mathrm{H}}^{\mathrm{H2O}} / N$, $N_{\mathrm{H2O}}$ is the number of atoms in water, and $N_{\mathrm{H}}^{\mathrm{H2O}}$ is the number of hydrogen atoms in water. 

**Jobs that deviate from the above mechanism**
 - RootMeanSquaredDeviation
RMSD doesn't follow the standard way in which the total results are summed up from individual components. The total results are not a sum or weighted sum of the individual components. RMSD results with grouping (e.g., `rmsd_[H2_O1]_all`) follow how the `rmsd_all` results are calculated, but with the atoms associated with a particular molecule. So `rmsd_[H2_O1]_all` will be the same as `rmsd_all` but with only `H2_O1` molecules selected.

- RootMeanSquaredFluctuation
Grouping for this job has been changed to `each atom` or `each molecule` just to make it clear that this job doesn't follow the same grouping mechanism explained above. The grouped results in this job are unchanged from what occurs on protos.

**Other jobs which could have a grouping**
- DensityProfile
I think grouping could be added to the DensityProfile job since it uses atom selection, atom transmutation, and weights. I have decided to leave it until #820 is resolved.
- Density
This job doesn't even split results into atomic parts yet, but I don't see why it couldn't be.

**All updated jobs**
- VanHoveFunctionDistinct
- DensityOfStates
- MeanSquaredDisplacement
- PositionAutoCorrelationFunction
- PositionPowerSpectrum
- VanHoveFunctionSelf
- VelocityAutoCorrelationFunction
- CurrentCorrelationFunction
- DynamicCoherentStructureFactor
- DynamicIncoherentStructureFactor
- ElasticIncoherentStructureFactor
- GaussianDynamicIncoherentStructureFactor
- StaticStructureFactorFromScatteringFunction
- NeutronTotalDynamicStructureFactor
- CoordinationNumber
- PairDistributionFunction
- StaticStructureFactor
- XRayStaticStructureFactor
- RootMeanSquaredDeviation
- RootMeanSquaredFluctuation

**Atom selection and transmutation**
Atom selection and transmutation work as before. With grouping atoms maintain their group after selection and transmutation so that if we have water molecules and select only the hydrogen atoms they will still be associated with the water molecule group. Similarly if we transmute the waters hydrogen atoms to deuterium or any other atoms, those atom will still be associated with the water molecule group.

**Example**
Here I have the DFTB water trajectory which I have transmuted some oxygen atoms to sulfur, so I have two different molecules, H2O (57 molecules) and H2S (43 molecules). 

<img width="319" alt="image" src="https://github.com/user-attachments/assets/52a7d3d7-8221-45da-a4b3-39102d0bcf77" />

Here I run DOS with molecule grouping. My total DOS result split into `dos_[H2_O1]_total` and  `dos_[H2_S1]_total` which get split into their atomic components, `dos_[H2_O1]_H` and `dos_[H2_O1]_O` for H2O and `dos_[H2_S1]_H` and `dos_[H2_S1]_S` for H2S. The total DOS is the sum of the total molecular DOS.

<img width="556" alt="image" src="https://github.com/user-attachments/assets/72598237-3a40-4b13-8715-9bcab35a6f4d" />

And the `dos_[H2_O1]_total` is a sum of its atomic components `dos_[H2_O1]_H` and `dos_[H2_O1]_O`.

<img width="554" alt="image" src="https://github.com/user-attachments/assets/2c200e06-8730-41c4-a452-4f0b09fa9c60" />

Here is the DISF for `f(q,t)_total`,`f(q,t)_[H2_O1]_total`, and `f(q,t)_[H2_S1]_total`.

<img width="616" alt="image" src="https://github.com/user-attachments/assets/259d08cc-b026-4732-a878-769417aa174e" />

From the explanation we had

```math
 F_{\mathrm{inc}, \mathrm{tot}}(q, t) = c_{\mathrm{H2O}}F_{\mathrm{inc}, \mathrm{H2O}}(q, t) +c_{\mathrm{EtOH}} F_{\mathrm{inc}, \mathrm{EtOH}}(q, t) 
```

We can also plot $F_{\mathrm{inc}, \mathrm{H2O}}(q, t)$ and $F_{\mathrm{inc}, \mathrm{H2S}}(q, t)$ instead of $c_{\mathrm{H2O}}F_{\mathrm{inc}, \mathrm{H2O}}(q, t)$ and $c_{\mathrm{EtOH}} F_{\mathrm{inc}, \mathrm{EtOH}}(q, t)$ by deselecting the `apply weight?` in the plotter to get

<img width="602" alt="image" src="https://github.com/user-attachments/assets/fc81255a-c602-403f-80b9-f22b1e31af9a" />

which gives you a result similar to a calculation where you selected those molecules (they are not exactly the same because the selection can change the weights normalisation factor #438)

Here is an example from the DOS. In the orange plot, only water molecules were selected and atom grouping was used. The blue is the total H2O results where I had all atoms selected with molecule grouping.

<img width="736" alt="image" src="https://github.com/user-attachments/assets/0c66610c-8ca6-45f8-881d-0fb78248befb" />

Now, if I normalise the data using the plotter, you can see they are the same; it's just that MDANSE's weight depends on the selection (#438).

<img width="733" alt="image" src="https://github.com/user-attachments/assets/fe17edc7-e197-4d1e-ba69-ec36c732d732" />

**To test**
Run a few jobs with atom selection, transmutation and grouping for trajectories with molecules and check that results follow expectations.

ChiCheng45 committed:
new grouping initial commit

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
update to allow selection, grouping, and transmutation

ChiCheng45 committed:
fix jobs

ChiCheng45 committed:
fix jobs

ChiCheng45 committed:
add grouping to DOS

ChiCheng45 committed:
edit

ChiCheng45 committed:
added grouping to PACF, PPS, self-part van Hove and VACF

ChiCheng45 committed:
applied ruff

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -72,73 +73,63 @@ def configure(self, value):
+                        if x not in atomSelectionConfig["flatten_indices"]:
+                            continue
+                        indices.append([x])
+                        flatten_indices.append(x)
```
I think this can be simplified, `more_itertools` provides the [`collapse`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.collapse) iterator which should iterate over the flattened indices.

Since it is subsequently stored, you might just be able to do
```python
flattened_indices = list(collapse(indices))
```

ChiCheng45 committed:
update grouped results names and refactored code

ChiCheng45 committed:
sorted sets in grouping config

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
updated weight functions and allowed group for DCSF

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
fix formatting

ChiCheng45 committed:
fixed docstrings

ChiCheng45 committed:
Merge branch 'protos' into chi/molecule-grouping

# Conflicts:
#	MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py

ChiCheng45 committed:
refactored code

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
fixed ssf

ChiCheng45 committed:
fixed ccf

ChiCheng45 committed:
updated distinct part of van hove function so that has grouping

ChiCheng45 committed:
updated distinct part of van hove function so that has grouping. fixed conc_exp for grouped total DCSF results

ChiCheng45 committed:
Updated so that CCF, DISF, EISF and GDISF so that they used grouping. Fixed bug in grouping when molecules are not selected.

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
cleanup

ChiCheng45 committed:
added grouping to SFFSF and NDTSF jobs.

ChiCheng45 committed:
added grouping to coordination number job.

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
fixed coordination number and change reference result as output results keys were updated.

ChiCheng45 committed:
added grouping to PDF job and refactored code

ChiCheng45 committed:
fixed docstring and applied ruff

ChiCheng45 committed:
fixed docstrings

ChiCheng45 committed:
fixed type hints

ChiCheng45 committed:
fixed type hints

ChiCheng45 committed:
edit

ChiCheng45 committed:
fixed pdf unit test reference result

ChiCheng45 committed:
fixed docstring

ChiCheng45 committed:
added grouping to SSF and fix grouping in PDF

ChiCheng45 committed:
fixed main/partial results for distinct van hove function

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
Merge branch 'protos' into chi/molecule-grouping

# Conflicts:
#	MDANSE/Src/MDANSE/Framework/Jobs/DensityProfile.py
#	MDANSE/Src/MDANSE/Framework/Jobs/VanHoveFunctionDistinct.py
#	MDANSE/Src/MDANSE/Mathematics/Arithmetic.py

ChiCheng45 committed:
added grouping to x-ray static structure factor and refactored code

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
fixed typing

ChiCheng45 committed:
fixed typing

ChiCheng45 committed:
fixed grouping for RMSD job

ChiCheng45 committed:
refactored code

ChiCheng45 committed:
add grouping level back to RMSF

ChiCheng45 committed:
updated grouping leve docstring

ChiCheng45 committed:
fixed GDISF grouping and added grouping regression tests

ChiCheng45 committed:
update test to pass on mac

ChiCheng45 committed:
update test to pass on mac

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -72,18 +81,59 @@ def configure(self, value):
         atomSelectionConfig = self._configurable[self._dependencies["atom_selection"]]
         chemical_system = trajConfig["instance"].chemical_system
+
+        if value == "atom" or value == "each atom":
```
```suggestion
        if value in {"atom", "each atom"}:
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -72,18 +81,59 @@ def configure(self, value):
+            self["group_elements"] = group_elements
+            self["group_n_atms"] = group_n_atms
+
+        if value == "each molecule":
```
If they're mutually exclusive:
```suggestion
        elif value == "each molecule":
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -72,18 +81,59 @@ def configure(self, value):
+                self.error_status = "The trajectory does not contain molecules."
+                return
+
+            for mol_name in chemical_system._clusters.keys():
```
If a `dict` `keys` is default iterator.
```suggestion
            for mol_name in chemical_system._clusters:
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -72,18 +81,59 @@ def configure(self, value):
+
+            self["name_to_element"] = {}
+            for name, element in zip(names, elements):
+                self["name_to_element"][name] = element[0]
```
```suggestion
            self["name_to_element"] = {name: element[0]
                                       for name, element in zip(names, elements)}
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
-                            counter += 1
 
         atomSelectionConfig["indices"] = indices
+        atomSelectionConfig["flatten_indices"] = collapse(indices)
```
Note this is an iterator and will exhaust on multiple uses. Probably want to still make this a list or tuple.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+        data_type: str,
+        dim: int = 1,
+        conc_exp: float = 1.0,
+        intra: bool = False,
```
`bool`s and large argument counts of extra arguments are usually best expressed as kwarg only. 
```suggestion
        *,
        intra: bool = False,
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+            raise NotImplementedError("Grouped total for dim > 2 not implemented.")
+
+    def pair_labels(
+        self, intra=False, all_pairs=False
```
```suggestion
        self, *, intra: bool = False, all_pairs: bool = False
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+
+    def update_pair_results(
+        self,
+        calc_func: Callable[[str, str], Iterator[tuple[str, bool, npt.NDArray]]],
```
To be precise, calc_func only needs to return an `Iterable` (can be used in a for-loop, i.e. has `__iter__`) rather than an `Iterator` (supports `next`)
```suggestion
        calc_func: Callable[[str, str], Iterable[tuple[str, bool, npt.NDArray]]],
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+                iterable = it.combinations_with_replacement(eles_i, 2)
+            else:
+                iterable = it.product(eles_i, eles_j)
+            for ele_i, ele_j in sorted(iterable):
```
In general the `product` or `combinations` of a sorted set will be sorted. This materialises the iterable and has to store it, which may be very large.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+                    else:
+                        output_data[f"{name}_[{grp_i}][{grp_j}]_{ele_i}{ele_j}"][
+                            ...
+                        ] = result
```
```suggestion
                    if intra and grp_i != grp_j:
                        continue

                    if intra and grp_i == grp_j:
                        output_data[f"{name}_[{grp_i}]_{ele_i}{ele_j}"][...] = result
                    else:
                        output_data[f"{name}_[{grp_i}][{grp_j}]_{ele_i}{ele_j}"][
                            ...
                        ] = result
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py:
```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+                            ...
+                        ] = result
+
+    def label_pairs(self, labels: list[str], all_pairs) -> list[tuple[str, str]]:
```
```suggestion
    def label_pairs(self, labels: Iterable[str], *, all_pairs: bool) -> list[tuple[str, str]]:
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/CoordinationNumber.py:
```diff
@@ -169,68 +144,69 @@ def finalize(self):
-            else:
-                nij = ni * nj
+        # symmetrize the data
+        for i, j in sorted(
```
Again, if these are pre-sorted, combinations will be.

MBartkowiakSTFC commented:
I am not sure if this should be the result:
1. I take a trajectory with 4 molecules (2 types):
![Screenshot 2025-06-09 at 11 48 41](https://github.com/user-attachments/assets/bdcdf6e2-f38b-44b3-a852-9c43437c43da)
2. I select a single molecule:
![Screenshot 2025-06-09 at 11 48 53](https://github.com/user-attachments/assets/8e0ec957-58d1-40fe-adab-67e80224d671)
3. The VanHoveFunctionDistinct with molecule grouping on this selection shows that inter- and intramolecular results are both non-zero.
![Screenshot 2025-06-09 at 11 50 04](https://github.com/user-attachments/assets/f56cb52f-c542-41ab-8b31-6679497e9305)

When the selection consists of a single molecule, I would expect the "inter" part to be 0.

ChiCheng45 commented:
I think the above issue is not related to this PR. There is something wrong with the trajectory or how the inter- or intra-molecular components are worked out. I've used the above trajectory and calculated the PDF with only the C60 selected on protos (8947373572069c1f3baaedfa744afa09eb810c39 because the plotter doesn't work on the latest commit).

<img width="269" alt="image" src="https://github.com/user-attachments/assets/d78d13c9-8985-44cf-bb6f-ae0074215af0" />

Running the PDF shows non-zero intra- and intermolecular parts.

<img width="646" alt="image" src="https://github.com/user-attachments/assets/40dba138-0094-4bc2-b4bb-b91c8b2415cd" />

oerc0122 commented on MDANSE/Src/MDANSE/Mathematics/Arithmetic.py:
```diff
@@ -125,20 +124,20 @@ def weighted_sum(
     """
-    weightedSum = 0.0
-    matches = {key % k for k in weights if k not in ["sum"]}
+    matches = [match_key % val[0] for val in match_labels]
```
Set contains should be faster.
```suggestion
    matches = {match_key % val[0] for val in match_labels}
```

oerc0122 commented on MDANSE/Src/MDANSE/Mathematics/Arithmetic.py:
```diff
@@ -77,7 +78,8 @@ def get_weights(
     weights: dict[str, float],
-    key: str,
+    match_key: str,
+    match_labels: Iterable,
```
```suggestion
    match_labels: Iterable[tuple[str, ?]],
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/XRayStaticStructureFactor.py:
```diff
@@ -173,79 +227,79 @@ def finalize(self):
+            for name, ele in zip(
+                self.configuration["atom_selection"]["names"],
+                self.configuration["atom_selection"]["elements"],
+            )
```
```python
    asf = {name: atomic_scattering_factor(
               ele[0],
               self._outputData["q"],
               self.configuration["trajectory"]["instance"],
           )
           for name, ele in zip(
                self.configuration["atom_selection"]["names"],
                self.configuration["atom_selection"]["elements"],
            )
        }

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/XRayStaticStructureFactor.py:
```diff
@@ -130,41 +140,85 @@ def finalize(self):
+
+            Yields
+            ------
+            tuple[str, bool, np.ndarray]
+                A tuple of the results name, a bool specifying whether
+                results correspond to intermolecular atom pairs, and
+                the results.
```
Numpydoc allows for:
```suggestion
            name : str
                Results name
            inter : bool
                Whether results are intermolecular atom pairs.
            results : np.ndarray
                The results.
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/PositionAutoCorrelationFunction.py:
```diff
@@ -90,6 +89,11 @@ def initialize(self):
 
         self.numberOfSteps = self.configuration["atom_selection"]["selection_length"]
 
+        self.labels = [
+            (element, (element,))
+            for element in self.configuration["atom_selection"].get_natoms()
+        ]
```
Seeing this a lot, does this just want to become a method where it returns 
```python
list(
    zip(
        self.configuration["atom_selection"].get_natoms(),
        combinations_with_replacement(
          
  self.configuration["atom_selection"].get_natoms(), 
            n=1
        )
    )
)
```
 for this style, n=2 for pairwise?

ChiCheng45 replied:
The n=2 case is more complicated; we have 

```python
    def pair_labels(
        self, *, intra: bool = False, all_pairs: bool = False
    ) -> list[tuple[str, tuple[str, str]]]:
```

in groupinglevelconfigurator for that. For the n=2 case, it needs to be corrected for intra-molecular labels and other things when required.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/RootMeanSquareDeviation.py:
```diff
@@ -150,20 +150,27 @@ def finalize(self):
 
+        nAtomsPerElement = self.configuration["atom_selection"].get_natoms()
         for element, number in nAtomsPerElement.items():
             self._outputData[f"rmsd_{element}"][:] = np.sqrt(
-                self._outputData[f"rmsd_{element}"]
+                self._outputData[f"rmsd_{element}"] / number
```
I realise it's in the original, but `number` is a little vague. `natoms`? `n_elem`?

ChiCheng45 replied:
This is done in a lot of places in MDANSE e.g.

https://github.com/ISISNeutronMuon/MDANSE/blob/93a56c72bd5c9bb17062fbc370a0256d3eab5251/MDANSE/Src/MDANSE/Framework/Jobs/DensityOfStates.py#L252-L254

so I think it's better to leave it, unless we decide to change it across the entire codebase. Also, I think it should be clear what `number` is referring to since it's iterating over `nAtomsPerElement.items()`.

oerc0122 reviewed (commented):
All looks reasonable and well tested, just a couple of stylistic things.

ChiCheng45 committed:
Apply suggestions from code review

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
Merge branch 'protos' into chi/molecule-grouping

ChiCheng45 committed:
Merge remote-tracking branch 'origin/chi/molecule-grouping' into chi/molecule-grouping

ChiCheng45 committed:
fixed type import

ChiCheng45 committed:
Merge branch 'protos' into chi/molecule-grouping

ChiCheng45 committed:
added all other suggestions

ChiCheng45 committed:
update other docstrings following suggestion

ChiCheng45 committed:
fixed DISF and GDISF grouped total results, fixed units

MBartkowiakSTFC commented on MDANSE/Src/MDANSE/Framework/Jobs/PositionPowerSpectrum.py:
```diff
@@ -246,25 +261,49 @@ def finalize(self):
+            "LineOutputVariable",
+            axis="time",
+            units="nm2",
+            main_result=True,
```
I think we should be more conservative with assigning the "main result" status to datasets. For PositionPowerSpectrum, I would not consider the position autocorrelation function to be the main result.

MBartkowiakSTFC commented:
Regarding the `main_result` and `partial_result` tags with molecule grouping, maybe we should say that:
- {something}_total is the **main** result
- {something}_{molecule}_total is **main** and **partial**

and for finer subdivisions the users will have to plot the results manually. I am just worried that the fast plot will become less convenient if it includes too many datasets.

ChiCheng45 committed:
fixed main and partial results

ChiCheng45 committed:
applied ruff

ChiCheng45 commented:
> Regarding the `main_result` and `partial_result` tags with molecule grouping, maybe we should say that:
> 
> * {something}_total is the **main** result
> * {something}_{molecule}_total is **main** and **partial**
> 
> and for finer subdivisions the users will have to plot the results manually. I am just worried that the fast plot will become less convenient if it includes too many datasets.

I've updated the code to follow the above. I also found that some datasets were set to a main results when they shouldn't have which I have now changed.

MBartkowiakSTFC reviewed (approved):
Thanks for the changes. I think this is fine now.
